registry/storage/metadata/metadata.go

```diff
@@ -0,0 +1,59 @@
+package metadata
+
+import (
+	"fmt"
+	"github.com/docker/distribution/context"
+	"github.com/docker/distribution/reference"
+)
+
+type ErrKeyNotFound struct {
+	Key string
+}
+
+func (err ErrKeyNotFound) Error() string {
+	return fmt.Sprintf("metadata key not found: %s", err.Key)
+}
+
+// Metadatable signifies metadata support
+type Metadatable interface {
+	MetadataService() MetadataService
+}
+
+type MetadataUpdateRecord struct {
+	Actual   interface{}
+	Expected interface{}
+}
+
+// IterFunc is called by Iterate.  A non-nil error stops iteration
+type IterFunc func(string, interface{}) error
+
+// MetadataService defines the operations of a metadata service
+type MetadataService interface {
+	// Put puts a value under a key
+	Put(context.Context, string, interface{}) error
+
+	// BatchPut performs a transactional put of multiple key value pairs
+	BatchPut(context.Context, map[string]MetadataUpdateRecord) error
+
+	// Get gets a value for a given key
+	Get(context.Context, string) (interface{}, error)
```

Comment:
In addition to the possibility of using an interface that has `Copy` defined, it definitely makes sense to use an interface that has some notion of equality. `reflect.DeepEqual` is slow and it won't always be the right way to compare two objects.